copyfromstaging.sh

```diff
@@ -0,0 +1,104 @@
+#!/bin/bash
+
+set -ex
+
+CHECK_COMPATIBILITY=true
+TMP=$(mktemp -d)
+
+copy_one_api() {
+  PACKAGE=$1
+  STAGING_DIR=owl-bot-staging/$1
+  PACKAGE_DIR=apis/$1
+
+  if [[ $CHECK_COMPATIBILITY == "true" && -d $PACKAGE_DIR ]]
+  then
+    echo "Building existing version of $PACKAGE for compatibility checking"
+    dotnet build -c Release -f netstandard2.0 -v quiet -nologo -clp:NoSummary -p:SourceLinkCreate=false $PACKAGE_DIR/$PACKAGE
+    cp $PACKAGE_DIR/$PACKAGE/bin/Release/netstandard2.0/$PACKAGE.dll $TMP
+  fi
+
+  # All source files need the extension .g.cs.  Some will only be generated with the
+  # extension .cs.  Rename files that end only in .cs.
+  for cs_file in $(find "$STAGING_DIR" -name '*.cs'); do
+    if [[ $cs_file != *.g.cs ]]; then
+      mv $cs_file $(echo $cs_file | sed -E 's/.cs$/.g.cs/g')
+    fi
+  done
+
+  # Remove the newly generated standalone snippets until they are ready for surfacing.
+  rm -rf "$STAGING_DIR/$1.StandaloneSnippets"
+
+  # We generate our own project files
+  rm -f $(find "$STAGING_DIR" -name '*.csproj')
+
+  # Wipe out all the existing source files we're about to replace.
+  # This will clear out source files that are no longer generated.
+  # BAD because this deleted GrpcCtorCompatibility.g.cs.
+  # rm -f $(find "$PACKAGE_DIR" -name '*.cs')
+
+  # Copy the files out of the staging directory into the package directory.
+  mkdir -p "$PACKAGE_DIR"
+  cp -rf $STAGING_DIR/* "$PACKAGE_DIR"
+
+  # Clean out the staging directory.
+  rm -rf "$STAGING_DIR"
+
+  # The following code was copied from generate_api().
```

Comment:
A note about pre/post-generation things in general...

We have some APIs where we modify the protos before generating, or between generating the DTOs/gRPC stubs and the GAPIC clients. I think those are going to be hard to migrate to OwlBot in the near term.

Anything with *just* a post-processing step as either a patch or a shell script should be fine - but I suspect there aren't many of those. I'll add another comment to this later with what's happening in each API where we have pre/post-processing, and how feasible OwlBot will be.

Reply:
Okay, here are the APIs with pre/post-processing. Note that I'm disregarding "undo the changes to the googleapis repo" (which is the most common post-processing step) as it wouldn't be relevant to OwlBot.

- Google.Cloud.AIPlatform.V1: Pregeneration makes two changes to the protos; one is somewhat C# specific and probably needs to stay. The other is a missing doc comment which should eventually be fixed. No post-processing.
- Google.Cloud.AccessApproval.V1: Mid/post-processing to rename the service "AccessApproval" to "AccessApprovalService". This is necessary as otherwise there's a message called "AccessApprovalSettings" which clashes with the GAPIC generated code.
- Google.Cloud.Bigtable.V2: Mid/post-processing to rename the service to `BigtableServiceApi`, then generate a `BigtableClient` which wraps it. (Bigtable is a bit of a snowflake API.)
- Google.Cloud.Compute.V1: Some DIREGAPIC-specific processing which should be able to go away in the longer term
- Google.Cloud.Dialogflow.V2: Pre-generation proto patching. In theory this could be removed by the API team applying the changes to the source protos, but I doubt that it'll happen. No post-generation.
- Google.Cloud.DocumentAI.V1: Pre-generation proto change to remove an unhelpful resource declaration. Removing this from the golden source proto would break other languages, unfortunately. No post-generation.
- Google.Cloud.DocumentAI.V1Beta3: Ditto
- Google.Cloud.GkeHub.V1: Fixes to csharp namespace options that are complicated to explain and hard to implement in the original source protos :( No post-generation
- Google.Cloud.Language.V1: Post-generation only fixes due to generated code conflicting with hand-written code.
- Google.Cloud.PubSub.V1: Mid-generation and post-generation changes to expose the "raw" services under a different name, so that PublisherClient and SubscriberClient are the handwritten ones.
- Google.LongRunning: Post-generation only, *but* it requires protoc and the original protos, so it may be hard to do with OwlBot
- Grafeas.V1: Post-generation only, fixing Grafeas (which is a slightly odd API as a sort of "interface without any host")

So out of these, I suspect that Grafeas.V1 and Google.Cloud.Language.V1 *can* be migrated to OwlBot easily, but the rest will take some more work.